DESIGN.md

```diff
@@ -0,0 +1,156 @@
+
+Dynamic dependencies are the opposite basically. Note that Nushell currently supports only static dependencies, but we might be able to add the "linking" feature at some point.
+
+We might want `nupm`support both types of dependencies.
+
+## Package repository 
+
+Packages need to be stored somewhere. There should be one central "official" location (see https://github.com/NixOS/nixpkgs for inspiration).
+
+Additionally, user should be able to add 3rd party repositories as well as install local and other packages (e.g., from the web, just pointing at URL),
+as long as it has `METADATA_FILE` telling `nupm` what to do.
+
+## API / CLI Interface
+
+Nushell's module design conflates CLI interface with API -- they are the same.
+
+- `nupm new [--script] [--module]`
+    - create a new local package with template files ([`kubouch/nuun`])
+- `nupm list`
+    - list currently installed packages and if they're out of date
+- `nupm install`
+    - install package into the currently active overlay (can override which overlay to install to)
+- `nupm add`
+    - add a dependency to the current project
+    - it is different from `nupm install`: this one adds the dependency to the MANIFEST_FILE, `nupm install` does not
+- `nupm uninstall`
+    - uninstall a package from a currently active overlay (can override which overlay to install to)
+- `nupm update`
+    - update all packages in a currently active overlay overlay (can specify package and/or overlay name)
+    - can be used to self-update (let's avoid having both "update" and "upgrade")
+- `nupm search`
+    - search package repository
+
+- `nupm check`
+    - parse the project to search for errors but do not run it
+- `nupm test`
+    - run unit and integration tests of local package
+- `nupm bench`
+    - run benchmarks
+- `nupm doc`
+    - generate documentation
+- `nupm publish`
+    - publish package to a repository
+    - (not sure about this one, for now, repository can be a github repo with packages submitted by PRs)
+
+- `nupm overlay new`
+    - create a new global overlay (Python's virtual environment style)
+    - `--local` flag could generate an overlay locally from the currently opened project
+- `nupm overlay remove`
+    - deletes the overlay
+- `nupm overlay list`
+    - list all overlays
+    - `nupm overlay list <overlay-name>` lists all packages installed within the overlay
+- `nupm overlay export`
+    - dump all the installed package names, versions, etc. to a file
+- `nupm overlay import`
+    - create overlay from exported file
+
+We could later think about being able to extend nupm, like cargo has plugins.
+
+## Other
+
+* activations
+* doc generation
+* test running
+* benchmark running
+* configuration (do not add until we really need something to be configurable, keep it minimal, case study of a project with minimal configuration: https://github.com/psf/black)
```

Comment:
Maybe something like that:

We will try to avoid configuration until it's really needed. When it will happen, we will most probably choose the idea of having `nupm` profiles.

Reply:
could you give an example of a "profile"? :yum:

Reply:
You mean what would we like to see specifically in `nupm`'s profile or just a real-life example from some package manager?

Reply:
maybe both?
i'm not familiar with this concept of "profile" :open_mouth:

Reply:
we can also keep that for later :thinking:

Reply:
I know profiles from [C(++) conan](https://docs.conan.io/1/reference/profiles.html)

It's basically a robust wrapping of config files.
```
> conan profile show myprofile1
Configuration for profile myprofile1:

[settings]
os=Windows
arch=x86_64
compiler=Visual Studio
compiler.version=15
build_type=Release
[options]
[tool_requires]
[env]
```

When it comes to the `nupm` I don't think we have something that needs configuration for now. My idea was: "If we will need some configuration, we will most probably use profiles". But if you feel it's too early for that, we can skip that part

Reply:
oooh i see

Reply:
so you'd have commands to
- create / delete profiles
- set configuration inside a profile
- switch between profiles
?

Reply:
i think we can skip that part for now, if that's ok with ya :relieved: 
the configuration part is mentionned at the end of the design, so we shouldn't forget about it :yum:

Reply:
> so you'd have commands to
> 
>     * create / delete profiles
> 
>     * set configuration inside a profile
> 
>     * switch between profiles
>       ?

Exactly. 



> i think we can skip that part for now, if that's ok with ya relieved the configuration part is mentionned at the end of the design, so we shouldn't forget about it yum

Fine, you can close this issue. I spend too much time with conan profesionally to forget about profiles 😅

Reply:
:+1: 

> I spend too much time with conan profesionally to forget about profiles
> 
:laughing: